src/vm/fastserializer.cpp

```diff
@@ -17,8 +17,8 @@ IpcStreamWriter::IpcStreamWriter(uint64_t id, IpcStream *pStream) : _pStream(pSt
     CONTRACTL
     {
         NOTHROW;
-        GC_TRIGGERS;
-        MODE_PREEMPTIVE;
+        GC_NOTRIGGER;
```

Comment:
This change is concerning, can you explain why this change is necessary?

Reply:
Agreed, and we should review our contracts.
This is the assertion I was getting without these changes:
```log
Assert failure(PID 16760 [0x00004178], Thread: 22220 [0x56cc]): CONTRACT VIOLATION by FastSerializer::WriteTag at "c:\coreclr\src\vm\fastserializer.cpp" @ 250

MODE_PREEMPTIVE encountered while thread is in cooperative state.

                        CONTRACT in FastSerializer::WriteTag at "c:\coreclr\src\vm\fastserializer.cpp" @ 250
                        CONTRACT in FastSerializer::WriteObject at "c:\coreclr\src\vm\fastserializer.cpp" @ 163
                        CONTRACT in EventPipeFile::Flush at "c:\coreclr\src\vm\eventpipefile.cpp" @ 117
                        CONTRACT in EventPipeBufferManager::WriteAllBuffersToFile at "c:\coreclr\src\vm\eventpipebuffermanager.cpp" @ 515
                        CONTRACT in EventPipeSession::WriteAllBuffersToFile at "c:\coreclr\src\vm\eventpipesession.cpp" @ 269
                        CONTRACT in ThreadProc at "c:\coreclr\src\vm\eventpipesession.cpp" @ 129

We can't find the violated contract. Look for an old-style non-holder-based contract.


CORECLR! CONTRACT_ASSERT + 0x345 (0x00007ffe`86f9a265)
CORECLR! EEContract::DoChecks + 0x2E6 (0x00007ffe`870ef826)
CORECLR! FastSerializer::WriteTag + 0x11B (0x00007ffe`87a7f5eb)
CORECLR! FastSerializer::WriteObject + 0x230 (0x00007ffe`87a7efc0)
CORECLR! EventPipeFile::Flush + 0x14B (0x00007ffe`87a8329b)
CORECLR! EventPipeBufferManager::WriteAllBuffersToFile + 0x609 (0x00007ffe`87a99539)
CORECLR! EventPipeSession::WriteAllBuffersToFile + 0x262 (0x00007ffe`878d0982)
CORECLR! <lambda_6ce8943d6e0712d6a69eb273347d9526>::operator() + 0x5D (0x00007ffe`878ceefd)
CORECLR! EventPipe::RunWithCallbackPostponed<<lambda_6ce8943d6e0712d6a69eb273347d9526> > + 0x46 (0x00007ffe`878ce3e6)
CORECLR! ThreadProc + 0x34C (0x00007ffe`878d06ac)
    File: c:\coreclr\src\vm\fastserializer.cpp Line: 250
    Image: C:\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe
```